aws/resource_aws_ebs_volume.go

```diff
@@ -373,3 +371,53 @@ func resourceAwsEbsVolumeDelete(d *schema.ResourceData, meta interface{}) error
 
 	return nil
 }
+
+func resourceAwsEbsVolumeCustomizeDiff(_ context.Context, diff *schema.ResourceDiff, meta interface{}) error {
+	iops := diff.Get("iops").(int)
+	multiAttachEnabled := diff.Get("multi_attach_enabled").(bool)
+	throughput := diff.Get("throughput").(int)
+	volumeType := diff.Get("type").(string)
+
+	if diff.Id() == "" {
+		// Create.
+
+		// Iops is required for io1 and io2 volumes.
+		// The default for gp3 volumes is 3,000 IOPS.
+		// This parameter is not supported for gp2, st1, sc1, or standard volumes.
+		// Hard validation in place to return an error if IOPs are provided
+		// for an unsupported storage type.
+		// Reference: https://github.com/hashicorp/terraform-provider-aws/issues/12667
+		switch volumeType {
+		case ec2.VolumeTypeIo1, ec2.VolumeTypeIo2:
+			if iops == 0 {
+				return fmt.Errorf("'iops' must be set when 'type' is '%s'", volumeType)
+			}
+
+		case ec2.VolumeTypeGp3:
+
+		default:
+			if iops != 0 {
+				return fmt.Errorf("'iops' must not be set when 'type' is '%s'", volumeType)
+			}
+		}
+
+		// MultiAttachEnabled is supported with io1 volumes only.
+		if multiAttachEnabled && volumeType != ec2.VolumeTypeIo1 {
+			return fmt.Errorf("'multi_attach_enabled' must not be set when 'type' is '%s'", volumeType)
+		}
+
+		// Throughput is valid only for gp3 volumes.
+		if throughput > 0 && volumeType != ec2.VolumeTypeGp3 {
+			return fmt.Errorf("'throughput' must not be set when 'type' is '%s'", volumeType)
+		}
```

Comment:
Just noting here that this type of validation can become a maintenance burden.

Reply:
I tend to agree but it's also no fun for practitioners to be hit with an error during `apply` that could have caught at `plan` or best `validate`.

Reply:
Of course, no one is arguing against that. The challenge here is that this type of information is not programmatically encoded in the API model and can differ per partition (or in worse cases per region within a partition), which means it needs manual curation should changes need to occur over time. It is no fun for practitioners to also be artificially limited in the future from doing a perfectly valid API call while waiting for this project to be updated. The most pragmatic answer probably lies somewhere in between both ends of the spectrum, adding it where there is a strong community ask and a strong maintainer ability to keep it up to date versus omitting it when either is not true.